monero-harness/src/lib.rs

```diff
@@ -44,219 +45,280 @@ const BLOCK_TIME_SECS: u64 = 1;
+        let monerod = self.inner();
+        let _ = tokio::spawn(mine(monerod, miner_wallet_address.to_string()));
         Ok(())
     }
+}
+
+impl<'c> MoneroWalletRpc {
+    /// Starts a new wallet container which is attached to
+    /// MONEROD_DEFAULT_NETWORK and MONEROD_DAEMON_CONTAINER_NAME
+    async fn new(
+        cli: &'c Cli,
+        name: &str,
+        monerod: &Monerod,
+    ) -> Result<(Self, Container<'c, Cli, image::Monero>)> {
+        let wallet_rpc_port: u16 =
+            port_check::free_local_port().ok_or_else(|| anyhow!("Could not retrieve free port"))?;
+
+        let daemon_address = format!("{}:{}", monerod.name, MONEROD_RPC_PORT);
+        let image = image::Monero::wallet(&name, daemon_address).with_mapped_port(Port {
+            local: wallet_rpc_port,
+            internal: WALLET_RPC_PORT,
+        });
+
+        let network = monerod.network.clone();
+        let run_args = RunArgs::default()
+            .with_name(name)
+            .with_network(network.clone());
+        let docker = cli.run_with_args(image, run_args);
+
+        // create new wallet
+        wallet::Client::localhost(wallet_rpc_port)
+            .create_wallet(name)
+            .await
+            .unwrap();
+
+        Ok((
+            Self {
+                rpc_port: wallet_rpc_port,
+                name: name.to_string(),
+                network,
+            },
+            docker,
+        ))
+    }
 
-    /// Get addresses for the primary account.
-    async fn get_address_miner(&self) -> Result<GetAddress> {
-        let wallet = self.miner_wallet_rpc_client();
-        wallet.get_address(ACCOUNT_INDEX_PRIMARY).await
+    pub fn inner(&self) -> wallet::Client {
+        wallet::Client::localhost(self.rpc_port)
     }
 
-    /// Get addresses for the Alice's account.
-    async fn get_address_alice(&self) -> Result<GetAddress> {
-        let wallet = self.alice_wallet_rpc_client();
-        wallet.get_address(ACCOUNT_INDEX_PRIMARY).await
+    // It takes a little while for the wallet to sync with monerod.
+    pub async fn wait_for_wallet_height(&self, height: u32) -> Result<()> {
+        let mut retry: u8 = 0;
+        while self.inner().block_height().await?.height < height {
+            if retry >= 30 {
+                // ~30 seconds
+                bail!("Wallet could not catch up with monerod after 30 retries.")
+            }
+            time::delay_for(Duration::from_millis(WAIT_WALLET_SYNC_MILLIS)).await;
+            retry += 1;
+        }
+        Ok(())
     }
 
-    /// Get addresses for the Bob's account.
-    async fn get_address_bob(&self) -> Result<GetAddress> {
-        let wallet = self.bob_wallet_rpc_client();
-        wallet.get_address(ACCOUNT_INDEX_PRIMARY).await
+    /// Sends amount to address
+    pub async fn transfer(&self, address: &str, amount: u64) -> Result<Transfer> {
+        let transfer = self.inner().transfer(0, amount, address).await?;
+        self.inner().refresh().await?;
```

Comment:
:question: Is this `refresh` necessary? From the [docs](https://web.getmonero.org/resources/developer-guides/wallet-rpc.html#refresh) it seems like it's only needed after opening the wallet. 

As a consumer of this library I only want to use the inner client and not have to keep around this `MoneroWalletRpc` struct. For example, in `xmr_btc` and `swap` we only ever use the underlying client.

Reply:
I found out that `refresh` triggers and active call to the monerod node retrieving all blocks. This should be faster than just `waiting_for_wallet_height` as monero-wallet-rpc polls monerod only every couple of seconds.

Reply:
Ah, so we could potentially get rid of `wait_for_wallet_height`. I still think it's not necessary to refresh the wallet after calling transfer, although maybe this has changed after separating the processes into different containers?

Reply:
Yes, that was actually wrong. The refresh is only be needed on the target wallet 🤦‍♀️.